Freezing Module containing fork subgraphs

bzinodev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38830 Preserve user specified attributes and methods
* **#37044 Freezing Module containing fork subgraphs**

This patch applies freezing to module containing fork subgraphs.

Differential Revision: [D21956334](https://our.internmc.facebook.com/intern/diff/D21956334)

bzinodev committed:
Freezing Module containing fork subgraphs

This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -332,12 +395,32 @@ class AttributePropagator {
               insertMutableAttr(name, attr, mptr);
               if (attr.isModule()) {
-                modules.insert(attr.toModule()._ivalue());
+                auto module = attr.toModule();
```
why is this change needed? if we insert the module and it already exists that should be fine

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -332,12 +395,32 @@ class AttributePropagator {
           }
+        } else if (n->kind() == prim::fork) {
+          applyToSubGraph(n, graph, [this](std::shared_ptr<Graph>& subgraph) {
+            recordReferencedAttrs(subgraph);
```
Why not just directly call `recordReferencedAttrs` here ?

bzinodev replied:
Fork is not called with  self as param. 
_fork(submodule, foo)

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -351,10 +434,14 @@ class AttributePropagator {
       return;
+    }*/
+    if (attrsToKeep_[type].count(N)) {
+      // Module escapes. Preserve all its attributes and methods.
```
What are these changes ?

eellison reviewed (commented):
Changes mostly look fine but it'd be easier to review with debug code removed

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -16,10 +16,33 @@ class AttributePropagator {
+        }
+        if (n->kind() == prim::fork) {
+          auto subgraph = n->g(attr::Subgraph);
+          optimizeSubGraphs(subgraph, func);
```
I ran performance on several and i do see regressions. We may need to run optimizations. I am looking into it to understand why there are regressions

eellison replied:
Those regressions should probably be handled separately. I think we should just run DCE - inlining will just adversely affect mobile.

bzinodev replied:
I submitted a separate patch

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -286,11 +318,42 @@ class AttributePropagator {
+      std::function<void(std::shared_ptr<Graph>&)> func) {
+    TORCH_CHECK(n->kind() == prim::fork);
+    auto attrModule = module_;
+    auto node = n->inputs()[0]->node();
```
It is a module (this in c++) but it is not enforced. 
This is how the attributes are accessed from fork subgraph

bzinodev replied:
Correct

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -286,11 +318,42 @@ class AttributePropagator {
+    auto attrModule = module_;
+    auto node = n->inputs()[0]->node();
+    // TODO: Short cut
+    TORCH_INTERNAL_ASSERT(node->kind() == prim::GetAttr);
```
Yes, that is correct. I changed the code to if ( ...=! GetAttr) return;

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -332,12 +395,32 @@ class AttributePropagator {
+          });
         }
+
+        /*if (n->kind() == prim::CallMethod) {
```
sorry, ghstack

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -64,13 +87,17 @@ class AttributePropagator {
     Node* node = input->node();
     names_.clear();
-    while (!(node->outputs()[0]->type() == attrModule.type())) {
+    while (!(node->outputs()[0]->type() == graph->inputs()[0]->type())) {
```
We use the first argument to  forward(self, ....) or fork(module, ...) to resolve  GetAttr chains

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

eellison reviewed (commented):
I think this also needs to add handling for the fork(free_function) case.

Edit: also, seems like this would not be that hard to generalize to the preserve multiple methods / attributes from ehere

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -286,11 +320,38 @@ class AttributePropagator {
           n->outputs().at(0)->replaceAllUsesWith(paramConst);
           n->removeAllInputs();
+        } else if (n->kind() == prim::fork) {
+          auto func = [this](std::shared_ptr<Graph>& subgraph) {
```
The lambda allows to capture 'this'.  Moreover, These are private methods so the compiler complains :). I think lambda wrap is neat.

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -286,11 +320,38 @@ class AttributePropagator {
+      return;
+    auto name = node->s(attr::name);
+    auto input = node->inputs()[0];
+    if (!findConstantAttr(input, name, attrModule, graph)) {
```
I think even if the module is a constant attribute we still need to run `recordMutableAttrs` on the subgraph here, so that modified attributes/submodules within the subgraph arrer preserved.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -286,11 +320,38 @@ class AttributePropagator {
+  void applyToSubGraph(
+      Node* n,
+      std::shared_ptr<Graph>& graph,
+      const std::function<void(std::shared_ptr<Graph>&)>& func) {
```
It might be easier to follow if this took in an enum of `recordMutableAttrs, recordReferencedAttrs, propagateAttributes` instead of a lambda

eellison reviewed (commented):
Looks good, I think we still need to make one change to run `recordMutableAttrs` on fork subgraph even if the parent module isn't constant

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

[ghstack-poisoned]

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

Differential Revision: [D21956334](https://our.internmc.facebook.com/intern/diff/D21956334)

[ghstack-poisoned]

bzinodev committed:
Update on "Freezing Module containing fork subgraphs"


This patch applies freezing to module containing fork subgraphs.

Differential Revision: [D21956334](https://our.internmc.facebook.com/intern/diff/D21956334)

[ghstack-poisoned]
